File: frontend/src/app/core/services/payment.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { InvoiceResponse, PayBillRequest, PaymentResponse } from '../models/payment.model';

@Injectable({ providedIn: 'root' })
export class PaymentService {
  private readonly baseUrl = '/api/payments';

  constructor(private readonly http: HttpClient) {}

  pay(request: PayBillRequest): Observable<PaymentResponse[]> {
    return this.http.post<PaymentResponse[]>(`${this.baseUrl}/pay`, request);
  }

  getInvoice(transactionId: string): Observable<InvoiceResponse> {
    return this.http.get<InvoiceResponse>(`${this.baseUrl}/invoice/${transactionId}`);
  }

  downloadReceipt(paymentId: string): Observable<Blob> {
    return this.http.get(`${this.baseUrl}/receipt/${paymentId}/download`, { responseType: 'blob' });
  }

  downloadInvoice(transactionId: string): Observable<Blob> {
    return this.http.get(`${this.baseUrl}/invoice/${transactionId}/download`, { responseType: 'blob' });
  }
}

File: backend/src/main/java/com/electricity/billing/service/PaymentService.java
package com.electricity.billing.service;

import com.electricity.billing.dto.request.PayBillRequest;
import com.electricity.billing.dto.response.InvoiceResponse;
import com.electricity.billing.dto.response.PaymentResponse;

import java.util.List;

public interface PaymentService {

    List<PaymentResponse> payBills(PayBillRequest request);

    InvoiceResponse getInvoice(String transactionId);

    byte[] getReceiptPdf(String paymentId);

    byte[] getInvoicePdf(String transactionId);
}
